src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs

```diff
@@ -100,12 +128,47 @@ public async Task AuthenticateAsync(IConnection connection, ConnectionDescriptio
             Ensure.IsNotNull(connection, nameof(connection));
             Ensure.IsNotNull(description, nameof(description));
 
-            using (var conversation = new SaslConversation(description.ConnectionId))
+            var firstStepResult = description.IsMasterResult.SpeculativeAuthenticate;
+
+            if (firstStepResult == null)
+            {
+                await AuthenticateAsync(
+                        connection, description, previousStep: null, previousStepResult: null, cancellationToken)
+                    .ConfigureAwait(false);
+            }
+            else
+            {
+                var (firstStep, _) = CreateFirstStepAndCommand(_mechanism, connection, conversation: null, description);
+                await AuthenticateAsync(
+                    connection,
+                    description,
+                    previousStep: firstStep,
+                    previousStepResult: firstStepResult,
+                    cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        private async Task AuthenticateAsync(
+            IConnection connection,
+            ConnectionDescription description,
+            ISaslStep previousStep,
+            BsonDocument previousStepResult,
+            CancellationToken cancellationToken)
+        {
+            Ensure.IsNotNull(connection, nameof(connection));
+            Ensure.IsNotNull(description, nameof(description));
+            if (previousStep != null)
             {
-                var currentStep = _mechanism.Initialize(connection, conversation, description);
+                Ensure.IsNotNull(previousStepResult, nameof(previousStepResult));
+            }
 
-                var command = CreateStartCommand(currentStep);
-                while (true)
+            using (var conversation = new SaslConversation(description.ConnectionId))
+            {
+                /* We cannot skip the cast to StepAndCommand without adding a reference to ValueTuple */
+                var (currentStep, command) = (StepAndCommand)(previousStep == null
```

Comment:
Same comments as the non-async version: https://github.com/vincentkam/mongo-csharp-driver/pull/95/files#r402537021